packages/chatbot-server-mongodb-public/src/tools/fetchPage.ts

```diff
@@ -0,0 +1,147 @@
+import { MakeReferenceLinksFunc } from "mongodb-chatbot-server";
+import { normalizeUrl } from "mongodb-rag-core/dataSources";
+import { wrapTraced } from "mongodb-rag-core/braintrust";
+import { Tool, tool, ToolExecutionOptions } from "mongodb-rag-core/aiSdk";
+
+export const FETCH_PAGE_TOOL_NAME = "fetch_page";
+export const SEARCH_ALL_FALLBACK_TEXT = "{fallback_to_search}";
+const SEARCH_ON_PAGE_LENGTH_CUTOFF = 150000;
+
+export const MongoDbFetchPageToolArgsSchema = z.object({
+  pageUrl: z.string().describe("URL of page to fetch"),
+  query: z.string().describe("Search query"),
+});
+
+export type MongoDbFetchPageToolArgs = z.infer<
+  typeof MongoDbFetchPageToolArgsSchema
+>;
+
+export type FetchPageToolResult = {
+  text: string;
+  references?: Reference[];
+};
+
+export type FetchPageTool = Tool<
+  typeof MongoDbFetchPageToolArgsSchema,
+  FetchPageToolResult
+> & {
+  execute: (
+    args: MongoDbFetchPageToolArgs,
+    options: ToolExecutionOptions
+  ) => PromiseLike<FetchPageToolResult>;
+};
+
+export interface MakeFetchPageToolParams {
+  loadPage: MongoDbPageStore["loadPage"];
+  findContent: FindContentFunc;
+  makeReferences: MakeReferenceLinksFunc;
+  searchFallbackText?: string;
+  pageLengthCutoff?: number;
+}
+
+export function makeFetchPageTool({
+  loadPage,
+  findContent,
+  makeReferences,
+  searchFallbackText = SEARCH_ALL_FALLBACK_TEXT,
+  pageLengthCutoff = SEARCH_ON_PAGE_LENGTH_CUTOFF,
+}: MakeFetchPageToolParams): FetchPageTool {
+  return tool({
+    parameters: MongoDbFetchPageToolArgsSchema,
+    description: "Fetch all content for a specific URL",
+    experimental_toToolResultContent(result) {
+      return [{ type: "text", text: result.text }];
+    },
+    execute: wrapTraced(
+      async function (
+        args: MongoDbFetchPageToolArgs,
+        _options: ToolExecutionOptions
+      ): Promise<FetchPageToolResult> {
+        const normalizedUrl = normalizeUrl({ url: args.pageUrl });
+        const page = await loadPage({
+          urls: [normalizedUrl],
+          query: {
+            action: {
+              $ne: "deleted",
+            },
+          },
+        });
+        return await getPageContent(
+          page,
+          pageLengthCutoff,
+          searchFallbackText,
+          findContent,
+          makeReferences,
+          args.query,
+          normalizedUrl
+        );
+      },
+      {
+        name: "fetchPageTool",
+      }
+    ),
+  });
+}
+
+async function getPageContent(
+  page: Awaited<ReturnType<MongoDbPageStore["loadPage"]>>,
+  pageLengthCutoff: number,
+  searchFallbackText: string,
+  findContent: FindContentFunc,
+  makeReferences: MakeReferenceLinksFunc,
+  query: string,
+  normalizedUrl: string
+): Promise<{
+  text: string;
+  references?: Reference[];
+}> {
+  if (page === null) {
+    // Fall back - no page for this URL
+    logger.info(
+      `${FETCH_PAGE_TOOL_NAME} did not find a page for URL ${normalizedUrl}, falling back to search tool`
```

Comment:
```suggestion
      `${FETCH_PAGE_TOOL_NAME} did not find a page for URL ${normalizedUrl}`
```

I think we should avoid coupling the `fetchPage` tool's implementation to the `search` tool. I could see a future where we don't always fall back to search and we'd need to remember to update this log too. Instead we can infer the fallback from the content of the logs (i.e. "fetchPage tool did not find a page" followed by "search tool found content" implies the fallback)